data-prepper-plugins/saas-source-plugins/saas-crawler/src/main/java/org/opensearch/dataprepper/plugins/source/saas_crawler/base/Crawler.java

```diff
@@ -0,0 +1,81 @@
+package org.opensearch.dataprepper.plugins.source.saas_crawler.base;
+
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.source.coordinator.enhanced.EnhancedSourceCoordinator;
+import org.opensearch.dataprepper.plugins.source.saas_crawler.coordination.partition.SaasSourcePartition;
+import org.opensearch.dataprepper.plugins.source.saas_crawler.coordination.state.SaasWorkerProgressState;
+import org.opensearch.dataprepper.plugins.source.saas_crawler.model.ItemInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Named;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Named
+public class Crawler {
+    private static final Logger log = LoggerFactory.getLogger(Crawler.class);
+    private static final int maxItemsPerPage = 20;
+    public static final String CREATED = "created";
+    public static final String UPDATED = "updated";
+
+    private final SaasClient client;
+
+    public Crawler(SaasClient client) {
+        this.client = client;
+    }
+
+    public long crawl(long lastPollTime,
+                      EnhancedSourceCoordinator coordinator) {
+        long startTime = System.currentTimeMillis();
+        client.setLastPollTime(lastPollTime);
+        Iterator<ItemInfo> itemInfoIterator = client.listItems();
+        log.info("Starting to crawl the source with lastPollTime: {}", lastPollTime);
+        long updatedPollTime = 0;
+        do {
+            final List<ItemInfo> itemInfoList = new ArrayList<>();
+            for (int i = 0; i < maxItemsPerPage && itemInfoIterator.hasNext(); i++) {
+                ItemInfo nextItem = itemInfoIterator.next();
+                if(nextItem==null) {
+                    //we don't expect null items, but just in case, we'll skip them
+                    continue;
+                }
+                itemInfoList.add(nextItem);
+                Map<String, Object> metadata = nextItem.getMetadata();
+                long niCreated = Long.parseLong(metadata.get(CREATED)!=null? (String)metadata.get(CREATED):"0");
```

Comment:
This is assuming that the metadata provides the values as `created` and `updated`, but that may vary between services.

See my comment below about making `ItemInfo` an interface. Then add two new methods:

```
Instant getCreatedAt();
Instant getLastUpdatedAt();
```

Each source can implement this in the way that makes the most sense for that source.

Reply:
Agree. Modified that class to be an interface and added just one method called `Instant getLastModifiedAt()`